### PR TITLE
VZ-2474: Platform operator network policies

### PR DESCRIPTION
**Pull request by fredtibbitts**
# Description

This PR adds the verrazzano-platform-operator network policies to the verrazzano-install namespace at platform operator container init time.

Implements VZ-2474

# Checklist 

As the author of this PR, I have:

- [x] Checked that I included or updated copyright and license notices in all files that I altered
- [x] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

**Review by bobdonat — APPROVED** [review]
Looks good.

File: platform-operator/internal/k8s/netpolicy.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	k8sAppLabel              = "k8s-app"
+	kubeDNSPodName           = "kube-dns"
+	nginxControllerPodName   = "ingress-controller"
+	appInstanceLabel         = "app.kubernetes.io/instance"
+	apiServerEndpointName    = "kubernetes"
+)
+
+// CreateOrUpdateNetworkPolicies creates or updates network policies for the platform operator to
+// limit network ingress and egress.
+func CreateOrUpdateNetworkPolicies(clientset kubernetes.Interface, client client.Client) (controllerutil.OperationResult, error) {
```
Comment (pfmackin) [inline platform-operator/internal/k8s/netpolicy.go:38]: This directory is supposed to be support functions that are generic.  Do we have an area where init container logic goes?  Maybe we need one

Comment (fredtibbitts) [reply platform-operator/internal/k8s/netpolicy.go]: I moved the netpolicy files out of the k8s package and put them in a new netpolicy package.

File: platform-operator/internal/netpolicy/netpolicy.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+	netv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	installNamespace      = "verrazzano-install"
```
Comment (pfmackin) [inline platform-operator/internal/netpolicy/netpolicy.go:20]: this should be in constants.go

File: platform-operator/internal/netpolicy/netpolicy.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+	if len(endpoints.Subsets) > 0 && len(endpoints.Subsets[0].Addresses) > 0 && len(endpoints.Subsets[0].Ports) > 0 {
+		return endpoints.Subsets[0].Addresses[0].IP, endpoints.Subsets[0].Ports[0].Port, nil
+	}
+
+	return "", 0, fmt.Errorf("unable to find a host and port for the kubernetes API server")
+}
+
+// makeNetworkPolicy returns a populated NetworkPolicy with ingress and egress rules for this operator.
+func makeNetworkPolicy(apiServerIP string, apiServerPort int32) *netv1.NetworkPolicy {
```
Comment (pfmackin) [inline platform-operator/internal/netpolicy/netpolicy.go:72]: this should be newNetworkPolicy

File: platform-operator/internal/netpolicy/netpolicy.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+// limit network ingress and egress.
+func CreateOrUpdateNetworkPolicies(clientset kubernetes.Interface, client client.Client) (controllerutil.OperationResult, error) {
+	ip, port, err := getAPIServerIPAndPort(clientset)
+	if err != nil {
+		return controllerutil.OperationResultNone, err
+	}
+
+	netPolicy := makeNetworkPolicy(ip, port)
+	// deep copy the spec because it can get overwritten when the network policy already exists
+	spec := netPolicy.Spec.DeepCopy()
```
Comment (pfmackin) [inline platform-operator/internal/netpolicy/netpolicy.go:46]: I don't think this is the normal way to use create or update.  I think you just pass in the object with namespace and name set then mutate (deeep copy included) in the func.  We have examples in our  code see multiclusterapplicationconfiguration/controller.go and see 
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controllerutil/example_test.go

File: platform-operator/internal/netpolicy/netpolicy.go (hunk 2 above)
Comment (fredtibbitts) [reply platform-operator/internal/netpolicy/netpolicy.go]: Done.

File: platform-operator/internal/netpolicy/netpolicy.go (hunk 3 above)
Comment (fredtibbitts) [reply platform-operator/internal/netpolicy/netpolicy.go]: Fixed.

File: platform-operator/internal/netpolicy/netpolicy.go (hunk 4 above)
Comment (fredtibbitts) [reply platform-operator/internal/netpolicy/netpolicy.go]: Fixed.

**Review by pfmackin — APPROVED** [review]

**Pull request merged**